src/Microsoft.TestPlatform.Common/ExtensionDecorators/SerializeTestRunDecorator.cs

```diff
@@ -0,0 +1,137 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Xml.Linq;
+
+using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging;
+
+namespace Microsoft.VisualStudio.TestPlatform.Common.ExtensionDecorators;
+
+internal class SerializeTestRunDecorator : ITestExecutor, ITestExecutor2, IDisposable
+{
+    private readonly SemaphoreSlim _runSequentialEvent = new(1);
+
+    public ITestExecutor OriginalTestExecutor { get; private set; }
+
+    public SerializeTestRunDecorator(ITestExecutor originalTestExecutor)
+    {
+        OriginalTestExecutor = originalTestExecutor;
+    }
+
+    public void RunTests(IEnumerable<TestCase>? tests, IRunContext? runContext, IFrameworkHandle? frameworkHandle)
+    {
+        if (IsSerializeTestRunEnabled(runContext))
+        {
+            EqtTrace.Info($"SerializeTestRunDecorator: Test cases will run sequentially");
+            if (tests is null)
+            {
+                return;
+            }
+
+            foreach (TestCase testToRun in tests)
+            {
+                _runSequentialEvent.Wait();
+                OriginalTestExecutor.RunTests(new List<TestCase> { testToRun }, runContext, new SerializeTestRunDecoratorFrameworkHandle(frameworkHandle!, _runSequentialEvent));
+            }
+        }
+        else
+        {
+            OriginalTestExecutor.RunTests(tests, runContext, frameworkHandle);
+        }
+    }
+
+    public void RunTests(IEnumerable<string>? sources, IRunContext? runContext, IFrameworkHandle? frameworkHandle)
+    {
+        if (IsSerializeTestRunEnabled(runContext))
+        {
+            EqtTrace.Error(Resources.Resources.SerializeTestRunInvalidScenario);
```

Comment:
I am a bit lost, are we localizing eqt messaged?

Reply:
This error will be shown in the user console but I wanted to have it inside the logs too, the line after this one is the message to the user.

Reply:
Is there any useful info/context we could capture as extra in the logs? Also it's probably better not to use translated version for logs, it would be hard to diagnose.